src/services/RecordStreamProcessor.ts

```diff
@@ -0,0 +1,110 @@
+import { ValidationHelper } from "../utils/ValidationHelper";
+import { personalIdentityUtils } from "../utils/PersonalIdentityUtils";
+import { Logger } from "@aws-lambda-powertools/logger";
+import { Metrics } from "@aws-lambda-powertools/metrics";
+import { ExtSessionEvent, SessionEvent } from "../models/SessionEvent";
+import { HttpCodesEnum } from "../models/enums/HttpCodesEnum";
+import { buildGovNotifyEventFields } from "../utils/GovNotifyEvent";
+import { EnvironmentVariables } from "./EnvironmentVariables";
+import { ServicesEnum } from "../models/enums/ServicesEnum";
+import { createDynamoDbClient } from "../utils/DynamoDBFactory";
+import { IPRService } from "./IPRService";
+import { AppError } from "../utils/AppError";
+import { MessageCodes } from "../models/enums/MessageCodes";
+import { Constants } from "../utils/Constants";
+
+export class RecordStreamProcessor {
+  private static instance: RecordStreamProcessor;
+  private readonly logger: Logger;
+  private readonly metrics: Metrics;
+  private readonly validationHelper: ValidationHelper;
+  private readonly iprService: IPRService;
+  private readonly environmentVariables: EnvironmentVariables;
+
+  constructor(logger: Logger, metrics: Metrics) {
+    this.logger = logger;
+    this.environmentVariables = new EnvironmentVariables(logger, ServicesEnum.RECORD_STREAM_PROCESSOR);
+    this.validationHelper = new ValidationHelper();
+    this.metrics = metrics;
+    this.iprService = IPRService.getInstance(this.environmentVariables.sessionEventsTable(), this.logger, createDynamoDbClient());
+  }
+
+  static getInstance(logger: Logger, metrics: Metrics): RecordStreamProcessor {
+    if (!RecordStreamProcessor.instance) {
+      RecordStreamProcessor.instance = new RecordStreamProcessor(logger, metrics);
+    }
+    return RecordStreamProcessor.instance;
+  }
+
+  async processRequest(sessionEvent: any): Promise<void> {
+
+    let sessionEventData: any = ExtSessionEvent.parseRequest(JSON.stringify(sessionEvent));
+
+    this.logger.appendKeys({ govuk_signin_journey_id: sessionEventData.clientSessionId });
+
+    if (sessionEventData.notified) {
+      this.handleNotifiedUser();
+    }
+
+    try {
+      this.validationHelper.validateSessionEventFields(sessionEventData);
+    } catch (error: any) {
+      this.handleValidationFailure(error);
```

Comment:
Again, is there a reason that we catch and then re-throw what is essentially the same error? Given we move the log into the `validateSessionEventFields` function